lib/github_test.go

```diff
@@ -0,0 +1,491 @@
+	for i := 1; i <= expectedSize; i++ {
+		team := "team" + strconv.Itoa(i)
+		if _, ok := mapTeamName[team]; !ok {
+			return fmt.Errorf("Team %v is missing", team)
+		}
+	}
+	return nil
+}
+
+func verifyAuthenticatedTokenReview(review *v1.TokenReview, teamSize int) error {
+	if !review.Status.Authenticated {
+		return fmt.Errorf("Expected authenticated ture, got false")
+	}
+	if review.Status.User.Username != "nahid" {
+		return fmt.Errorf("Expected username %v, got %v", "nahid", review.Status.User.Username)
+	}
+	if review.Status.User.UID != "1204" {
+		return fmt.Errorf("Expected user id %v, got %v", "1204", review.Status.User.UID)
+	}
+	err := verifyTeamList(review.Status.User.Groups, teamSize)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func verifyUnauthenticatedTokenReview(review *v1.TokenReview, expectedErr string) error {
+	if review.Status.Authenticated {
+		return fmt.Errorf("Expected authenticated false, got true")
+	}
+	if !strings.Contains(review.Status.Error, expectedErr) {
+		return fmt.Errorf("Expected error `%v`, got `%v`", expectedErr, review.Status.Error)
+	}
+	return nil
+}
+
+func githubServerSetup(githubOrg string, memberResp string, memberStatusCode int, genTeamRespn teamRespFunc) *httptest.Server {
+	m := pat.New()
+
+	m.Get(fmt.Sprintf("/user/memberships/orgs/%v", githubOrg), http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		err := verifyAuthorization(r)
+		if err != nil {
+			w.WriteHeader(http.StatusUnauthorized)
+			w.Write(getErrorMessage(err))
+			return
+		}
+		w.WriteHeader(memberStatusCode)
+		w.Write([]byte(memberResp))
+	}))
+
+	m.Get(fmt.Sprintf("/user/memberships/orgs/"), http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		w.WriteHeader(http.StatusUnauthorized)
+		w.Write(getErrorMessage(errors.New("Authorization: invalid token")))
+		return
+	}))
+
+	m.Get("/user/teams", http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		err := verifyAuthorization(r)
+		if err != nil {
+			w.WriteHeader(http.StatusUnauthorized)
+			w.Write(getErrorMessage(err))
+			return
+		}
+
+		status, resp := genTeamRespn(r.URL)
+		w.WriteHeader(status)
+		if status != http.StatusOK {
+			w.Write(getErrorMessage(errors.New(resp)))
+			return
+		}
+		w.Write([]byte(resp))
+	}))
+
+	srv := httptest.NewServer(m)
+	return srv
+}
+
+func githubClientSetup(serverUrl, githubOrg string, ctx context.Context, httpClient *http.Client) (*GithubClient, error) {
+	g := &GithubClient{
+		Ctx:     ctx,
+		OrgName: githubOrg,
+	}
+	var err error
+	g.Client, err = github.NewEnterpriseClient(serverUrl, serverUrl, httpClient)
+	if err != nil {
+		return nil, err
+	}
+	return g, nil
+}
+
+func TestCheckGithub(t *testing.T) {
+	//	ref:
+	//  	https://developer.github.com/v3/orgs/members/#get-your-organization-membership
+	var (
+		//user membership response body
+		memBody1 = `
+{
+   "user":{
+      "login":"nahid",
+      "id":1204
+   }
+}
+`
+	)
+
+	teamSizes := []int{25, 0, 1, 13, 100, 77, 233}
+
```

Comment:
Restructure the test data like https://github.com/kubernetes/kubernetes/pull/54366/files#diff-11ca6bae99dd41cf90ffc5528959802dR175 , so that the actual test code becomes simple